### PR TITLE
Add support for Teams Adaptive cards in QnA Dialog

### LINKED ISSUE #6630: QnAMakerDialog cannot render rich text in Teams channel
The following text is copied from an Internal Ticket:

The problem happens on the Azure Cognitive Service - Custom Question Answering exported Bot application.

[PROBLEM DESCRIPTION]
This is a UI rendering issue only happens on Teams channel.
I notice the CQA generated Bot application is including rich text (based on Markdown format), it can be rendered in Web Chat correctly.

In a Web Chat bot, markdown formatting is rendered correctly.

In a Teams chat bot, the formatting is not rendered correctly.

I digest into the Bot application generated by CQA, seems it's based on the BotBuilder samples - QnAMaker bot. I use Nodejs SDK but .NET SDK shows the same problem.

I tried to update all BotBuilder dependencies to latest 4.19.2 but still no improvements.
If this is a mis-usage of the QnAMakerDialog class, which class should be used to show the same layout as Web Chat?

### PR DESCRIPTION
**Pull request by anishprasad01**
Fixes Internal IcM Issue #373550043
Port of changes from https://github.com/microsoft/botbuilder-js/pull/4467
Fixes #6630 

## Description

This PR allows for choosing Adaptive Cards formatted for use in Microsoft Teams instead of the default Hero Cards. This is done to support markdown formatting in QnA and CQA responses since Teams only supports markdown in Adaptive Cards. Changes are made in `QnACardBuilder.cs` and `QnAMakerDialog.cs`.

Using the new cards simple requires adding code in the bot to take in a new environment variable: UseTeamsAdaptiveCard = "true" and pass it to the QnA dialog constructor. If no changes are made, the bot will default to a Hero card.

## Specific Changes

- Create new methods to generate adaptive and hero cards
- Modify the card builder class to take in a new bool field to allow for choosing whether or not to use an adaptive card
- Modify the dialog class to accept the aforementioned bool field and pass it to the card builder.

## Testing

Tests tun successfully.
Screenshot 2023-04-27 114013

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll**

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs [hunk 1]
```diff
@@ -126,39 +127,130 @@ public static IMessageActivity GetQnADefaultResponse(QueryResult result, BoolExp
...
-                    heroCardText = result.Answer;
+                    cardText = result.Answer;
                 }
             }
 
-            if (buttonList != null || !string.IsNullOrWhiteSpace(heroCardText))
+            if (buttonList != null || !string.IsNullOrWhiteSpace(cardText))
             {
-                var plCard = new HeroCard();
+                bool useAdaptive = useTeamsAdaptiveCard == null ? false : useTeamsAdaptiveCard.Value;
```
Comment (ceciliaavila) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs:144]:
```suggestion
                var useAdaptive = useTeamsAdaptiveCard == null ? false : useTeamsAdaptiveCard.Value;
```

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs [hunk 2]
```diff
@@ -126,39 +127,130 @@ public static IMessageActivity GetQnADefaultResponse(QueryResult result, BoolExp
...
+                chatActivity.Attachments.Add(cardAttachment);
+            }
+
+            return chatActivity;
+        }
+
+        /// <summary>
+        /// Get a Teams-formatted Adaptive Card as Attachment to be returned in the QnA response. Max width and height of response are controlled by Teams.
+        /// </summary>
+        /// <param name="cardText">string of text to be added to the card.</param>
```
Comment (ceciliaavila) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs:156]:
```suggestion
        /// <param name="cardText">String of text to be added to the card.</param>
```

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs [hunk 3]
```diff
@@ -126,39 +127,130 @@ public static IMessageActivity GetQnADefaultResponse(QueryResult result, BoolExp
...
+        /// <param name="buttonList">List of CardAction representing buttons to be added to the card.</param>
+        /// <returns>Attachment.</returns>
+        private static Attachment CreateAdaptiveCardAttachment(string cardText, List<CardAction> buttonList)
+        {
+            // If there are buttons, create an array of buttons for the card.
+            // Each button is represented by a Dictionary containing the required fields for each button.
+            var cardButtons = buttonList != null ? buttonList.Select(button =>
```
Comment (ceciliaavila) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs:163]:
```suggestion
            var cardButtons = buttonList?.Select(button =>
```

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs [hunk 4]
```diff
@@ -126,39 +127,130 @@ public static IMessageActivity GetQnADefaultResponse(QueryResult result, BoolExp
...
-                // Create the attachment.
-                var attachment = plCard.ToAttachment();
+            // If there are buttons, add the buttons array to the card. "actions" must be formatted as an array.
+            if (cardButtons != null)
+            {
+                card.Add("actions", cardButtons);
+            }
+
+            // Create and return the card as an attachment
+            Attachment adaptiveCard = new Attachment()
```
Comment (ceciliaavila) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/QnACardBuilder.cs:222]:
```suggestion
            var adaptiveCard = new Attachment()
```

**Review by ceciliaavila — CHANGES_REQUESTED** [review]
Just some small details. Also, we could add unit tests to cover the new methods.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll**

**Comment by anishprasad01** [comment]
@ceciliaavila,

Could you please provide feedback on the new tests?

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll**

**Review by ceciliaavila — APPROVED** [review]

**Review by tracyboehrer — APPROVED** [review]

**Pull request merged**